enterprise/server/byte_stream_server_proxy/byte_stream_server_proxy.go

```diff
@@ -6,92 +6,218 @@ import (
+func New(env environment.Env) (*ByteStreamServerProxy, error) {
+	local := env.GetLocalByteStreamClient()
+	if local == nil {
+		return nil, fmt.Errorf("A local ByteStreamClient is required to enable ByteStreamServerProxy")
 	}
-	remoteCache := env.GetByteStreamClient()
-	if remoteCache == nil {
-		return nil, fmt.Errorf("A ByteStreamClient is required to enable ByteStreamServerProxy")
+	remote := env.GetByteStreamClient()
+	if remote == nil {
+		return nil, fmt.Errorf("A remote ByteStreamClient is required to enable ByteStreamServerProxy")
 	}
 	return &ByteStreamServerProxy{
-		env:         env,
-		localCache:  localCache,
-		remoteCache: remoteCache,
+		env:    env,
+		local:  local,
+		remote: remote,
 	}, nil
 }
 
 func (s *ByteStreamServerProxy) Read(req *bspb.ReadRequest, stream bspb.ByteStream_ReadServer) error {
-	remoteStream, err := s.remoteCache.Read(stream.Context(), req)
+	localReadStream, err := s.local.Read(stream.Context(), req)
 	if err != nil {
-		return err
+		return s.readRemote(req, stream)
 	}
+	responseSent := false
 	for {
-		rsp, err := remoteStream.Recv()
+		rsp, err := localReadStream.Recv()
 		if err == io.EOF {
 			break
 		}
 		if err != nil {
+			// If some responses were streamed to the client, just return the
+			// error. Otherwise, fall-back to remote.
+			if responseSent {
+				return err
+			} else {
+				return s.readRemote(req, stream)
+			}
+		}
+
+		if err := stream.Send(rsp); err != nil {
+			return err
+		}
+		responseSent = true
+	}
+	return nil
+}
+
+// The Write() RPC requires the client keep track of some state. This struct
+// and its methods take care of that.
+type localWriter struct {
+	ctx          context.Context
+	local        bspb.ByteStream_WriteClient
+	resourceName string
+	initialized  bool
+	offset       int64
+}
+
+func (s *localWriter) send(data []byte) error {
+	req := &bspb.WriteRequest{WriteOffset: s.offset, Data: data}
+	if !s.initialized {
+		// Read resources have a different name format than written resources
+		// Re-write the resource name with a fake execution ID so it passes
+		// the write regex.
+		req.ResourceName = "/uploads/00000000-0000-0000-0000-000000000000" + s.resourceName
+		s.initialized = true
+	}
+	s.offset += int64(len(data))
+	return s.local.Send(req)
+}
+
+func (s *localWriter) close() error {
+	if err := s.local.Send(&bspb.WriteRequest{WriteOffset: s.offset, FinishWrite: true}); err != nil {
+		return err
+	}
+	// Ignore the local response (but not the error)
+	_, err := s.local.CloseAndRecv()
+	return err
+}
+
+func (s *ByteStreamServerProxy) readRemote(req *bspb.ReadRequest, stream bspb.ByteStream_ReadServer) error {
+	remoteReadStream, err := s.remote.Read(stream.Context(), req)
+	if err != nil {
+		return err
+	}
+
+	var localWriteStream *localWriter = nil
+	if req.ReadOffset == 0 {
+		localStream, err := s.local.Write(stream.Context())
+		if err != nil {
+			log.Warningf("error writing to local bytestream server: %s", err)
```

Comment:
did you mean to initialize localWriteStream with a nil localStream? I think that would lead to a panic.

Reply:
oh nice catch, fixed